[docs] Add CI/CD tutorial (#44866)
expo/expo: docs/pages/tutorial/cicd/tag-based-releases.mdx

---
title: 'Using Git tags to trigger production deployments'
sidebar_title: Tag-based releases
description: Learn how to trigger production deployments from Git tags on EAS Workflows.
---

import { ContentSpotlight } from '~/ui/components/ContentSpotlight';
import { BranchVsTagTrigger } from '~/ui/components/FlowDiagram/Diagrams/BranchVsTagTrigger';
import { ProgressTracker } from '~/ui/components/ProgressTracker';
import { Terminal } from '~/ui/components/Snippet';
import { Step } from '~/ui/components/Step';

> Pushing to a release branch runs the production workflow on every commit, whether the team meant that commit as a release or not.

Reviewer: Two tweaks: "whether the team meant" assumes a team (this applies solo too), and the opener has no resolution. Suggest:

```suggestion
Pushing to a release branch runs the production workflow on every commit, even ones we never meant as a release. In this chapter, we switch the trigger to a Git tag, so production runs only when we push a release tag.
```

Author: Applied the solo-inclusive rewording ("even ones we never meant as a release"). Skipped the added resolution sentence for the same reason as the other chapter openers: the Learning outcomes list directly below carries the chapter's payoff.

## Learning outcomes

- Switch the production workflow trigger from a `release/*` branch to a Git tag matching `v*.*.*`
- Cut a release by pushing a tag from `main` and watch the workflow run on the EAS dashboard
- Match pre-release tags like `v*.*.*-rc.*` to run the workflow on release candidates without shipping them to app users

## Tags as release events

> EAS Workflows supports tag-based triggers through [`on.push`](/eas/workflows/syntax/#onpush) with a `tags` list. The workflow runs only when a tag matching a glob pattern is pushed to the remote. Unlike a branch that can be updated over time, a Git tag is a one-time event tied to a specific commit and version, which makes it well-suited for marking releases.

Author: **[suggestion]** Last sentence is ~30 words and stacks a contrast clause, an apposition, and a result clause (`expo-docs-style-guide.md#voice-and-tone`).

```suggestion
EAS Workflows supports tag-based triggers through [`on.push`](/eas/workflows/syntax/#onpush) with a `tags` list. The workflow runs only when a tag matching a glob pattern is pushed to the remote. A Git tag is a one-time event tied to a specific commit and version, unlike a branch that can be updated over time. That makes tags well-suited for marking releases.
```

A tag-based release workflow fits teams that:

> - Want the version number in Git to match the version shipped to app users

Author: **[suggestion]** Passive `shipped to app users` (`expo-docs-style-guide.md#voice-and-tone`).

```suggestion
- Want the version number in Git to match the version they ship to app users
```

- Release from `main` directly rather than maintaining long-lived release branches
- Want a clear audit trail of every release commit by version

<BranchVsTagTrigger />

## Switch production.yml to a tag trigger

> Open **.eas/workflows/production.yml** and replace the `on.push.branches` trigger with `on.push.tags`. Everything else inside the workflow stays the same.

Reviewer: **production.yml** comes from the previous chapter, but readers who land here directly (tag-based releases is a common direct-entry topic) have no file and no pointer to where it came from. Naming the origin keeps the chapter standalone:

```suggestion
Open the **.eas/workflows/production.yml** workflow from the [previous chapter](/tutorial/cicd/production/) and replace the `on.push.branches` trigger with `on.push.tags`. Everything else inside the workflow stays the same.
```

    type: get-build
    params:
      fingerprint_hash: ${{ needs.fingerprint.outputs.ios_fingerprint_hash }}
      profile: production
  build_android:
    name: Build Android
    needs: [get_android_build]
    if: ${{ !needs.get_android_build.outputs.build_id }}
    type: build
    params:
      platform: android
      profile: production
  build_ios:
    name: Build iOS
    needs: [get_ios_build]
    if: ${{ !needs.get_ios_build.outputs.build_id }}
    type: build
    params:
      platform: ios
      profile: production
  update_android:
    name: Publish Android update
    needs: [get_android_build]
    if: ${{ needs.get_android_build.outputs.build_id }}
    type: update
    params:
      branch: production
      platform: android
  update_ios:
    name: Publish iOS update
    needs: [get_ios_build]
    if: ${{ needs.get_ios_build.outputs.build_id }}
    type: update
    params:
      branch: production
      platform: ios
```

For tags to trigger the workflow, we need to pick a pattern that matches our team's flow. Common patterns include:

- `['v*']` matches any tag that starts with `v`
- `['v*.*.*']` matches strict three-part semantic versions (recommended)

> No other job changes are needed. The `fingerprint`, `get-build`, `build`, and `update` jobs work identically because they operate on the commit.

Author: **[suggestion]** Passive voice (`expo-docs-style-guide.md#voice-and-tone`).

```suggestion
We don't need to change any other jobs. The `fingerprint`, `get-build`, `build`, and `update` jobs work identically because they operate on the commit.
```

    '',
    '$ git push origin main',
  ]}
  cmdCopy="git add .eas/workflows/production.yml && git commit -m 'Switch production workflow to tag trigger' && git push origin main"
/>

</Step>

<Step label="3">

### Create a tag to test the workflow

To run the workflow, we need to push a tag that matches the glob pattern defined in **production.yml**.

Make a TypeScript/JavaScript change in our example project, commit it, and push a new tag:

<Terminal
  cmd={[
    '$ git add .',
    '',
    '$ git commit -m "Fix welcome copy"',
    '',
    '$ git tag v0.1.0',
    '',
    '$ git push origin main --tags',
  ]}
  cmdCopy="git add . && git commit -m 'Fix welcome copy' && git tag v0.1.0 && git push origin main --tags"
/>

The workflow triggers on the `v0.1.0` tag. Open the EAS dashboard, and under **Workflows**, notice "Deploy to production" running with `refs/tags/v0.1.0` as its branch:

<ContentSpotlight
  alt="EAS Workflows dashboard showing a tag-triggered production workflow run with fingerprint, get-build, and update jobs for Android and iOS."
  src="/static/images/tutorial/cicd/eas-workflows-tag-based-production-run.png"
  className="max-w-160"
/>

> Since this is the first production run with the fingerprint generated for a tag-based release, the `build` jobs run for both Android and iOS. Subsequent tags that point to TypeScript/JavaScript changes will skip the build jobs and go straight to `update` jobs.

Author: **[nit]** Present tense (`expo-docs-style-guide.md#voice-and-tone`).

```suggestion
Since this is the first production run with the fingerprint generated for a tag-based release, the `build` jobs run for both Android and iOS. Subsequent tags that point to TypeScript/JavaScript changes skip the build jobs and go straight to `update` jobs.
```

<ContentSpotlight
  alt="EAS Workflows dashboard showing parallel Android and iOS production build jobs for a v0.1.0 tag release."
  src="/static/images/tutorial/cicd/eas-workflows-tag-based-build-android-ios.png"
  className="max-w-160"
/>

</Step>

## Pre-release tags for release candidates

> To run the production workflow for release candidates, use a separate glob that matches pre-release tags. For example, `['v*.*.*-rc.*']` matches `v1.0.0-rc.1`.

Reviewer: This section may under-deliver on learning outcome 3 ("run the workflow on release candidates **without shipping them to app users**"). Two things I couldn't resolve as a reader:

1. **The "without shipping" mechanism isn't shown.** If an `rc` tag triggers this same production workflow, the `update` jobs still publish to users - matching a pre-release glob by itself doesn't gate anything. Does the intended pattern use a separate workflow for rc tags (e.g. build/internal-test but skip the production update), or something else?
2. **"a separate glob" is ambiguous** - separate workflow file, or a second pattern in the same `tags` list? And depending on glob semantics, would `v*.*.*` also match `v1.0.0-rc.1` (so an rc tag fires the real production workflow too)?

Happy to help expand this section if you can share the intended pattern - alternatively, if rc gating is out of scope for the tutorial, trimming outcome 3 to match would resolve the mismatch.

Author: You were right on both counts: `v*.*.*` does match `v1.0.0-rc.1`, so as written an rc tag would have fired the real production workflow, and the section never showed a gating mechanism. Rewrote the section: the production trigger now excludes pre-release tags with an exclusion glob (`tags: ['v*.*.*', '!v*.*.*-rc.*']`), and the text explains running candidates in a separate workflow that reuses the build jobs without the update and submit jobs. Learning outcome 3 is reworded to match. Sharpest catch of the review, thank you.
